### PR TITLE
Add culture dependant object formatting in REPL

### PR DESCRIPTION
**Pull request by drognanar**
When the REPL displays the output information it gets presented in the
culture of the running submission.

Handles issue #7133

### CODE REVIEW

File: src/Scripting/CSharpTest/CommandLineRunnerTests.cs [hunk 1]
```diff
@@ -82,6 +82,35 @@ public void Await()
...
+[en-GB]
+> Math.PI
+3.1415926535897931
+> DefaultThreadCurrentCulture = GetCultureInfo(""de-DE"")
+[de-DE]
+> Math.PI
+3,1415926535897931
+>", runner.Console.Out.ToString());
+        }
+
```
Comment (davkean) [inline src/Scripting/CSharpTest/CommandLineRunnerTests.cs]: I'd like to see additional tests here;

1) What if I set the culture via Thread.CurrentCulture = new CultureInfo("de-DE"), is this respected?
2) What if I fire off a background thread, does the culture correctly flow onto it, for example, `await Task.Run(() => { return 10.ToString() };`.

Comment (drognanar) [inline src/Scripting/CSharpTest/CommandLineRunnerTests.cs]: I added a test to check that a Task respects a `DefaultThreadCurrentCulture`
However setting `Thread.CurrentCulture` is not respected due to a separate bug #7132

**Comment by drognanar** [comment]
FYI @dotnet/roslyn-interactive @amcasey

File: src/Compilers/CSharp/Portable/SymbolDisplay/ObjectDisplay.cs [hunk 2]
```diff
@@ -346,7 +346,7 @@ internal static string FormatLiteral(sbyte value, ObjectDisplayOptions options)
             }
             else
             {
-                return value.ToString(CultureInfo.InvariantCulture);
+                return value.ToString(FormatCulture(options));
```
Comment (amcasey) [inline src/Compilers/CSharp/Portable/SymbolDisplay/ObjectDisplay.cs]: Consider something like "GetFormatCulture".  This way it sounds like it's the culture that's being formatted.

File: src/Compilers/CSharp/Test/Symbol/SymbolDisplay/ObjectDisplayTests.cs [hunk 3]
```diff
@@ -203,16 +203,19 @@ public void CultureInvariance()
                 var decimalValue = new Decimal(12.5);
                 Assert.Equal("12,5", decimalValue.ToString());
                 Assert.Equal("12.5", ObjectDisplay.FormatLiteral(decimalValue, ObjectDisplayOptions.None));
+                Assert.Equal("12,5", ObjectDisplay.FormatLiteral(decimalValue, ObjectDisplayOptions.UseCurrentCulture));
```
Comment (amcasey) [inline src/Compilers/CSharp/Test/Symbol/SymbolDisplay/ObjectDisplayTests.cs]: Can you please add a test that passes both `UseCurrentCulture` and `UseHexadecimal`?  It looks like hex ignores the current culture.  I'll tentatively assume that that's by design, but we should probably test/document it.

File: src/Scripting/Core/Hosting/ObjectFormatter/ObjectFormatter.cs [hunk 4]
```diff
@@ -531,7 +531,9 @@ internal virtual string FormatMethodSignature(MethodBase method)
 
         internal static ObjectDisplayOptions GetObjectDisplayOptions(bool useHexadecimalNumbers)
         {
-            return useHexadecimalNumbers ? ObjectDisplayOptions.UseHexadecimalNumbers : ObjectDisplayOptions.None;
```
Comment (amcasey) [inline src/Scripting/Core/Hosting/ObjectFormatter/ObjectFormatter.cs]: Can you please add ObjectFormatter tests documenting this behavior?

File: src/Scripting/VisualBasicTest/CommandLineRunnerTests.vb [hunk 5]
```diff
@@ -129,6 +129,32 @@ Type ""#help"" for more information.
 >", runner.Console.Out.ToString())
         End Sub
 
+        Public Sub TestDisplayResults()
```
Comment (amcasey) [inline src/Scripting/VisualBasicTest/CommandLineRunnerTests.vb]: `<Fact>` and `<WorkItem>`?

File: src/Scripting/Core/Hosting/ObjectFormatter/ObjectFormatter.cs (hunk 4 above)
Comment (amcasey) [inline src/Scripting/Core/Hosting/ObjectFormatter/ObjectFormatter.cs]: I see that this is covered by the command line tests.  Fair enough.

File: src/Scripting/CSharpTest/CommandLineRunnerTests.cs [hunk 6]
```diff
@@ -82,6 +82,41 @@ public void Await()
         }
 
         [Fact]
+        public void DisplayResults()
```
Comment (amcasey) [inline src/Scripting/CSharpTest/CommandLineRunnerTests.cs]: `[WorkItem]`?

**Comment by amcasey** [comment]
FYI @KevinH-MS, who has a related work item.

File: src/Compilers/CSharp/Portable/SymbolDisplay/ObjectDisplay.cs [hunk 7]
```diff
@@ -505,11 +505,18 @@ internal static string FormatLiteral(decimal value, ObjectDisplayOptions options
...
-            var result = value.ToString(CultureInfo.InvariantCulture);
+            var result = value.ToString(FormatCulture(options));
 
             return options.IncludesOption(ObjectDisplayOptions.IncludeTypeSuffix) ? result + "M" : result;
         }
 
+        private static CultureInfo FormatCulture(ObjectDisplayOptions options)
+        {
+            return options.IncludesOption(ObjectDisplayOptions.UseCurrentCulture)
+                ? CultureInfo.CurrentCulture
```
Comment (KevinH-MS) [inline src/Compilers/CSharp/Portable/SymbolDisplay/ObjectDisplay.cs]: I believe this should use CurrentUICulture, not CurrentCulture.

File: src/Scripting/CSharpTest/CommandLineRunnerTests.cs (hunk 6 above)
Comment (KevinH-MS) [inline src/Scripting/CSharpTest/CommandLineRunnerTests.cs]: Naming (perhaps something along the lines of DisplayResultsWithCurrentUICulture?)

File: src/Compilers/CSharp/Portable/SymbolDisplay/ObjectDisplay.cs (hunk 7 above)
Comment (KevinH-MS) [inline src/Compilers/CSharp/Portable/SymbolDisplay/ObjectDisplay.cs]: (and if we change this, we should change the name of the property)

**Comment by dpoeschl** [comment]
Note: the failure in prtest/win/dbg/unit64 is known and tracked as https://github.com/dotnet/roslyn/issues/8077

File: src/Compilers/CSharp/Portable/SymbolDisplay/ObjectDisplay.cs [hunk 8]
```diff
@@ -483,33 +483,38 @@ internal static string FormatLiteral(ulong value, ObjectDisplayOptions options)
...
             ValidateOptions(options);
 
-            var result = value.ToString("R", CultureInfo.InvariantCulture);
+            var result = value.ToString("R", GetFormatCulture(cultureInfo));
 
             return options.IncludesOption(ObjectDisplayOptions.IncludeTypeSuffix) ? result + "F" : result;
         }
 
-        internal static string FormatLiteral(decimal value, ObjectDisplayOptions options)
+        internal static string FormatLiteral(decimal value, ObjectDisplayOptions options, CultureInfo cultureInfo = null)
```
Comment (genlu) [inline src/Compilers/CSharp/Portable/SymbolDisplay/ObjectDisplay.cs]: Maybe have a default value of `CultureInfo.InvariantCulture` instead and save a method invocation and comparasion down the line.

File: src/Compilers/VisualBasic/Portable/SymbolDisplay/ObjectDisplay.vb [hunk 9]
```diff
@@ -303,26 +303,26 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.ObjectDisplay
...
             ValidateOptions(options)
 
-            Dim result = value.ToString("R", CultureInfo.InvariantCulture)
+            Dim result = value.ToString("R", GetFormatCulture(cultureInfo))
 
             Return If(options.IncludesOption(ObjectDisplayOptions.IncludeTypeSuffix), result & "F", result)
         End Function
 
-        Friend Function FormatLiteral(value As Decimal, options As ObjectDisplayOptions) As String
+        Friend Function FormatLiteral(value As Decimal, options As ObjectDisplayOptions, Optional cultureInfo As CultureInfo = Nothing) As String
```
Comment (genlu) [inline src/Compilers/VisualBasic/Portable/SymbolDisplay/ObjectDisplay.vb]: Default to `CultureInfo.InvariantCulture`?

**Comment by genlu** [comment]
LGTM other than my comments on default argument value.

File: src/Compilers/CSharp/Portable/SymbolDisplay/ObjectDisplay.cs (hunk 8 above)
Comment (drognanar) [inline src/Compilers/CSharp/Portable/SymbolDisplay/ObjectDisplay.cs]: Default value must be a compile time constant which `CultureInfo.InvariantCulture` is not.
The alternatives would thus be to inline all cals to `GetFormatCulture` or overload all `FormatLiteral` methods.

Comment (genlu) [inline src/Compilers/CSharp/Portable/SymbolDisplay/ObjectDisplay.cs]: Ah, I see. I thought it was a static field. Please ignore my comments then.

**Comment by drognanar** [comment]
PR update

File: src/Scripting/Core/Hosting/ObjectFormatter/ObjectFormatter.cs [hunk 10]
```diff
@@ -12,6 +12,7 @@
 
 namespace Microsoft.CodeAnalysis.Scripting.Hosting
 {
+    using System.Globalization;
```
Comment (amcasey) [inline src/Scripting/Core/Hosting/ObjectFormatter/ObjectFormatter.cs]: Can you please move this out of the namespace declaration?  The alias is only in here to resolve an ambiguity (I assume).

**Comment by amcasey** [comment]
I look forward to merging with this. :)

**Comment by amcasey** [comment]
:+1:

**Pull request merged**